synapse/handlers/federation.py

```diff
@@ -629,7 +633,8 @@ async def do_invite_join(
                 state_events=state,
             )
 
-            if ret.partial_state:
+            already_partially_joined = await self.store.is_partial_state_room(room_id)
```

Comment:
There's still a race:
1. We enter `do_invite_join()`.
2. The room is un-partial stated in the background.
3. We read the partial state flag and see that it's not present.
4. We then proceed to re-partial state the room (!) which breaks a key assumption.

I propose we introduce a lock that guarantees that no other code may toggle the partial stated-ness of the room while the lock is held. That is, when joining:
1. We take the lock for the room
2. We do a partial send-join when i) we aren't in the room or ii) the room is partial stated. Corollary: we do a full join when we are in the room and it's not partial stated. This is the same condition as mentioned in another comment.
3. We persist the join, which should no longer be able to raise a `PartialStateConflictError`
4. We release the lock right after the code that kicks off the sync.

When syncing, we take the lock when we're about to finish and un-partial state the room.

We can use https://github.com/matrix-org/synapse/blob/develop/synapse/storage/databases/main/lock.py for the locking, but it needs an `acquire()` method to be implemented (keep attempting `try_acquire()` and sleeping? not ideal but implementing it properly with replication is a fair bit of work)

TODO: check in #synapse-devs if this makes sense.

Reply:
@DMRobertson Heads up that some of the above may impact leaves while partial stated. I'm not entirely sure yet.

Reply:
> Maybe the partial state flag should imply that we are currently trying to sync partial state

FWIW this was my internal model for the meaning of that flag.

I'm not quite following all the moving changes here. I might have to pick your brains tomorrow

Reply:
> Maybe the partial state flag should imply that we are currently trying to sync partial state

We now think that the partial state flag just means that there are partial state events in the room and does not imply that there is a partial state sync running (because the partial state sync may have failed). It also does not imply that the local user currently has a member in the room.

Reply:
The proposed locking was implemented in #14882.